app/lib/service/rate_limit/rate_limit.dart

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:clock/clock.dart';
+import 'package:collection/collection.dart';
+import 'package:pub_dev/audit/backend.dart';
+
+import '../../account/agent.dart';
+import '../../audit/models.dart';
+import '../../shared/configuration.dart';
+import '../../shared/exceptions.dart';
+import '../../shared/redis_cache.dart';
+
+/// Verifies if the current package upload has a rate limit and throws
+/// if the limit has been exceeded.
+Future<void> verifyPackageUploadRateLimit({
+  required AuthenticatedAgent agent,
+  required String package,
+}) async {
+  final operation = AuditLogRecordKind.packagePublished;
+  if (agent.email != null) {
+    await _verifyRateLimit(
+      rateLimit: _getRateLimit(operation, RateLimitScope.user),
+      dataFilters: {'email': agent.email!},
```

Comment:
Why not filter on the user-id ?

Reply:
service accounts share the same agent id, but have different email that we can filter on

Reply:
Updated to hopefully use this logic. PTAL.